### ISSUE TITLE
ClassEmbedding  为什么需要 drop_mask = (-1 * (1 - drop_mask))  # need to flip 0 <-> 1

### ISSUE THREAD

**Issue by 1197680865** [FutureXiang/Diffusion#2]
大佬好，我对ClassEmbedding  中drop_mask的操作有些疑问，想深入请教下，望解答：
[转到 drop_mask 翻转操作](https://github.com/FutureXiang/Diffusion/blob/dfa611cb696900d450d3629638d54019a402c844/model/block.py#L58)
1. ClassEmbedding中为什么要对drop_mask 翻转，这个操作与One-hot有关系么？ 
2. 如果我的Embedding不是one-hot类型的（比如类似embedding这种连续变量的），我需要如何使用drop_mask的翻转操作呢？

谢谢

**Comment by FutureXiang** [comment]
你好。
Classifier-Free Guidance提出在训练时随机drop掉`drop_prob=10%`的condition，训练一个conditional和unconditional的混合模型：
https://github.com/FutureXiang/Diffusion/blob/dfa611cb696900d450d3629638d54019a402c844/model/DDPM_guide.py#L76-L77
此处的condition并不一定是label，还可以是text embedding, source image embedding, ... 等。

如果是one-hot label，则drop的可行方法是将one-hot label乘零、得到全零label，定义全零label为unconditional。
翻转的原因是上面取训练样本的时候是以`drop_prob`概率取1、以`1-drop_prob`概率取0。但我们希望 `drop_mask==1`对应zero-hot vector； `drop_mask==0`对应one-hot vector。
https://github.com/FutureXiang/Diffusion/blob/dfa611cb696900d450d3629638d54019a402c844/model/block.py#L58-L59

如果是其他离散或连续embedding，实践时仍然可以定义全零（或某个特殊状态）为drop掉、无信息，比如：
```py
class DropEmbedding(nn.Module):
    def __init__(self, in_channel, out_channel):
        super().__init__()
        self.lin1 = nn.Linear(in_channel, out_channel)
        self.act = nn.SiLU()
        self.lin2 = nn.Linear(out_channel, out_channel)
    
    def forward(self, emb, drop_mask):
        drop_mask = drop_mask[:, None]
        drop_mask = drop_mask.repeat(1, self.n_classes)
        drop_mask = (-1 * (1 - drop_mask))  # need to flip 0 <-> 1
        emb = emb * drop_mask

        emb = self.act(self.lin1(c))
        emb = self.lin2(emb)
        return emb
```

**Comment by 1197680865** [comment]
您的解答很详细，感谢👍！
另外我还是对下面这行代码有些疑惑，从您的解答以及代码注释来看是实现`0转1、1转0`，但是我debug了下输出是`0转-1、1转0`，所以有些疑惑，可能是我理解错了，或者对Python语法理解有误，还是想请有时间了您解释下，感谢

https://github.com/FutureXiang/Diffusion/blob/dfa611cb696900d450d3629638d54019a402c844/model/block.py#L58

以下是运行结果：
```python
c = torch.randn(2,5)
mask = torch.bernoulli(torch.zeros_like(c) + 0.1)
#tensor([[0., 1., 0., 0., 0.],
#       [1., 0., 0., 1., 0.]])

mask = (-1 * (1 - mask))
#tensor([[-1., -0., -1., -1., -1.],
#       [-0., -1., -1., -0., -1.]])
# 这里mask翻转后0转-1、1转0了
```

**Comment by FutureXiang** [comment]
这段代码沿用自 https://github.com/TeaPearce/Conditional_Diffusion_MNIST ，理论上相当于整个embedding取了负，我认为实际效果应该没有差别，因此没有改动。
至于为什么原作者取了$-1$，也有人提过[issue](https://github.com/TeaPearce/Conditional_Diffusion_MNIST/issues/1#issue-1336736505)，但是原作者没有回答。我个人不认为有特殊的好处。

**Comment by 1197680865** [comment]
谢谢解答。看来应该是Conditional_Diffusion_MNIST的作者在这里用来什么技巧吧，我用`(-1 * (1 - drop_mask))` 和 `(1 - drop_mask)`两种都试验了下，从结果来看 `(-1 * (1 - drop_mask))` 效果更好。
谢谢大佬啦！
